### PR TITLE
feat(lidar_centerpoint): remove libtorch from centerpoint

### PR DESCRIPTION
**Pull request by yukke42**
## Related Issue(required)

https://github.com/autowarefoundation/autoware.universe/issues/488

## Description(required)

- remove libtorch and implement the equivalent libraries for this package using c++/cuda
- change converted onnx model format, but the pre-trained model is the same as before
- update links to download trained models

before:
- GPU memory: 2817MiB
- inference time: 48.79 ± 3.78 ms

after:
- GPU memory: 1295MiB
- inference time: 45.43 ± 3.37 ms

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [commit-guidelines][commit-guidelines]
- [x] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by yukkysaito** [comment]
@yukke42 pls add related PR

**Comment by yukke42** [comment]
@yukkysaito 
libtorch isn't installed in https://github.com/autowarefoundation/autoware

**Comment by yukkysaito** [comment]
@yukke42 cc @kenji-miyake 
how are developpers doing without libtorch?

Does this mean that detection cannot work in https://github.com/autowarefoundation/autoware?

**Comment by yukke42** [comment]
@yukkysaito 
We used cuda 11.1 and tensorrt 7.2.1, but https://github.com/autowarefoundation/autoware requires cuda 11.4 and tensorrt 8.2.2. ([here](https://github.com/autowarefoundation/autoware/blob/main/ansible/playbooks/universe.yaml)). Perception packages which requires CUDA  aren't supported yet.  So, we should make sure that lidar_centerpoint and other packages, s.t.  lidar_apollo_instance_segmentation, tensorrt_yolo, ..etc, work on both environment at the same time.

**Comment by yukke42** [comment]
@yukkysaito @aohsato @mkuri It's now ready for review.

**Comment by mkuri** [comment]
@miursh ?

**Comment by miursh** [comment]
I checked it's working in my pc.
before:
GPU memory: 2819MiB
inference time: 31.94 ± 6.92 ms

after:
GPU memory: 1299MiB
inference time: 27.23 ± 6.94 ms

**Comment by kaancolak** [comment]
I tested it in my computer with sample-map-rosbag and it's working well. 

Currently, documentation of [rosbag replay simulation](https://autowarefoundation.github.io/autoware-documentation/main/tutorials/ad-hoc-simulation/rosbag-replay-simulation/) not working properly.

**Comment by yukke42** [comment]
> I tested it in my computer with sample-map-rosbag and it's working well.
> 
> Currently, documentation of [rosbag replay simulation](https://autowarefoundation.github.io/autoware-documentation/main/tutorials/ad-hoc-simulation/rosbag-replay-simulation/) not working properly.

@kaancolak Thank you for testing this PR.
Cloud you create an issue in https://github.com/autowarefoundation/autoware-documentation and describe the problem in more detail?

**Comment by yukke42** [comment]
`build-and-test-differential / build-and-test-differential (pull_request) ` wasn't passed because the model failed to be downloaded from gdrive.

Screenshot from 2022-03-23 09-59-35

**Comment by yukke42** [comment]
> `build-and-test-differential / build-and-test-differential (pull_request) ` wasn't passed because the model failed to be downloaded from gdrive.
> 
> https://github.com/autowarefoundation/autoware.universe/runs/5640881412?check_suite_focus=true
> 
> Screenshot from 2022-03-23 09-59-35

I will remove to check downloaded files to pass ci.

**Comment by miursh** [comment]
@wep21 Since I'm not familiar with CUDA programming, could you help me reviewing this PR?

File: perception/lidar_centerpoint/lib/include/nms_kernel.hpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2022 TIER IV, Inc.
```
Comment (wep21) [inline perception/lidar_centerpoint/lib/include/nms_kernel.hpp:1]: [nits] rename to `circle_nms_kernel.hpp`

File: perception/lidar_centerpoint/lib/src/nms_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,142 @@
+// Copyright 2022 TIER IV, Inc.
```
Comment (wep21) [inline perception/lidar_centerpoint/lib/src/nms_kernel.cu:1]: [nits] rename to circle_nms_kernel.cu

File: perception/lidar_centerpoint/lib/src/nms_kernel.cu [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  const Box3D * boxes, const std::size_t num_boxes3d, const std::size_t col_blocks,
+  const float dist2d_pow_threshold, std::uint64_t * mask)
+{
+  // params: boxes (N,)
+  // params: mask (N, divup(N/THREADS_PER_BLOCK_NMS))
+
+  const auto row_start = blockIdx.y;
+  const auto col_start = blockIdx.x;
+
+  if (row_start > col_start) return;
```
Comment (wep21) [inline perception/lidar_centerpoint/lib/src/nms_kernel.cu:52]: This line is commented out in original code.
Why do you need it?
https://github.com/open-mmlab/OpenPCDet/blob/a35f429768b3aa6ea2e7cd0a6452e1fcb7671c8d/pcdet/ops/iou3d_nms/src/iou3d_nms_kernel.cu#L336

File: perception/lidar_centerpoint/lib/include/nms_kernel.hpp (hunk 1 above)
Comment (yukke42) [reply perception/lidar_centerpoint/lib/include/nms_kernel.hpp]: changed at [b108e57](https://github.com/autowarefoundation/autoware.universe/pull/480/commits/b108e5746ba83aa2b01241b16dd99aca1edda90c)

File: perception/lidar_centerpoint/lib/src/nms_kernel.cu (hunk 2 above)
Comment (yukke42) [reply perception/lidar_centerpoint/lib/src/nms_kernel.cu]: changed at [b108e57](https://github.com/autowarefoundation/autoware.universe/pull/480/commits/b108e5746ba83aa2b01241b16dd99aca1edda90c)

**Comment by wep21** [comment]
@miursh I have taken a look at the part of cuda implementation, I guess there is no problem.

**Review by wep21 — APPROVED** [review]
LGTM

**Pull request merged**